liblouis/utils.c

```diff
@@ -276,3 +323,92 @@ _lou_isValidMode(int mode) {
 		mode &= ~validTranslationModes[i];
 	return !mode;
 }
+
+/* Map char to dots according to North American Braille Computer Code (NABCC) */
+widechar EXPORT_CALL
+_lou_charToFallbackDots(widechar c) {
+	static const unsigned char charToDots[] = {
+		/* ASCII characters 0X00-0X1F - control characters.
+		 * These won't be referenced so we have room for data.
+		 * These groups must be in descending order.
+		 * Each group contains the following four bytes:
+		 * 1) The first character to which this block applies.
+		 * 2) The bits to remove from the character.
+		 * 3) The bits to add to the character.
+		 * 4) The dots to add to the braille pattern.
+		 */
+		0X7F, 0X20, 0X00, LOU_DOT_7, 0X60, 0X20, 0X00, 0, 0X5F, 0X00, 0X00, 0, 0X40, 0X00,
+		0X00, LOU_DOT_7, 0X20, 0X00, 0X00, 0, 0X00, 0X00, 0X40, LOU_DOT_7 | LOU_DOT_8,
+
+		// ASCII characters 0X20-0X3F - digits and common symbols.
+		[' '] = 0, ['!'] = LOU_DOT_2 | LOU_DOT_3 | LOU_DOT_4 | LOU_DOT_6,
```

Comment:
I do not understand the `[' '] = 0` notation. What value is that putting into the array?

Reply:
The thing in brackets is just the index within the array. The part after the equals sign is the value at that index.

What does make it a bit difficult to read is that clang-format messed up the layout. I wonder if there's anything we can do to prevent it from doing that.

Reply:
> I wonder if there's anything we can do to prevent clang-format from messing up the layout.

The [doc](https://clang.llvm.org/docs/ClangFormatStyleOptions.html#disabling-formatting-on-a-piece-of-code) says:

``` C
int formatted_code;
// clang-format off
    void    unformatted_code  ;
// clang-format on
void formatted_code_again;
```

Reply:
Yes, OK, that's not really what I meant but we could indeed use this.